### PR TITLE
Retrieve stored calibration data

### PR DESCRIPTION
**Pull request by MFernandezCarmona**
Hi,
I needed to retrieve the calibration data from the sensors in order to parse raw data packets, so I added that functionality to the library. I tried to follow the general style, but please feel free to correct anything you don't like.
I also added a couple of files for users to test the library, but I'm not sure if those are useful.

Cheers,

### CODE REVIEW

**Comment by lumagi** [comment]
Hi Manuel,

thanks a ton for your contribution! I will probably take a closer look at the code over the weekend.

Cheers,
Lukas

File: pyshimmer/dev/calibration.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        reg_bin_str = fmt_hex(self._reg_bin)
+        obj_str += f'Binary: {reg_bin_str}\n'
+
+        return obj_str
+
+    def __eq__(self, other: "AllCalibration") -> bool:
+        return self._reg_bin == other._reg_bin
+
+    def check_sens_num(self,sens_num: int) -> None:
```
Comment (lumagi) [inline pyshimmer/dev/calibration.py:56]: Please make this protected as well

File: pyshimmer/dev/calibration.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+import re
```
Comment (lumagi) [inline pyshimmer/dev/calibration.py:16]: Can you use some basic PEP8 formatter to format the file? I wrote most of the code in PyCharm, they have some decent basic formatting support. I know that I don't have any clear specification what should be used and tbh yours is the biggest contribution to the project I have had so far.

File: pyshimmer/dev/calibration.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        obj_str += f'Binary: {reg_bin_str}\n'
+
+        return obj_str
+
+    def __eq__(self, other: "AllCalibration") -> bool:
+        return self._reg_bin == other._reg_bin
+
+    def check_sens_num(self,sens_num: int) -> None:
+        if not 0 <= sens_num <= (self.num_sensors-1):
+            raise ValueError('Sensor num must be 0 to ' + (self.num_sensors-1))
```
Comment (lumagi) [inline pyshimmer/dev/calibration.py:58]: Please also use f-strings here

File: pyshimmer/dev/calibration.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        reg_bin_str = fmt_hex(self._reg_bin)
+        obj_str += f'Binary: {reg_bin_str}\n'
+
+        return obj_str
+
+    def __eq__(self, other: "AllCalibration") -> bool:
+        return self._reg_bin == other._reg_bin
+
+    def check_sens_num(self,sens_num: int) -> None:
+        if not 0 <= sens_num <= (self.num_sensors-1):
```
Comment (lumagi) [inline pyshimmer/dev/calibration.py:57]:
```suggestion
        if not 0 <= sens_num < self.num_sensors:
```

File: pyshimmer/dev/calibration.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            raise ValueError('Sensor num must be 0 to ' + (self.num_sensors-1))
+
+    def get_offset_bias(self, sens_num: int) -> List[int]:
+        self.check_sens_num(sens_num)
```
Comment (lumagi) [inline pyshimmer/dev/calibration.py:61]: I don't think you really need this loop. You can simply tell struct that you want to unpack a total of three 16bit values. Then you get a tuple of length 3 that already contains all your values. If you want, you can cast it to a list with `list()`. This comment applies to the `get_sensitivity` method as well.

File: pyshimmer/dev/calibration.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        ans = list()
+        offset = sens_num * self.sensor_bytes + 6
+        for num_i in range(0,3):
+            d_i = struct.unpack('>h', self._reg_bin[offset:offset+2])
+            offset = offset + 2
+            ans.append(d_i[0])
+        return ans
+
+    def get_ali_mat(self, sens_num: int) -> List[int]:
```
Comment (lumagi) [inline pyshimmer/dev/calibration.py:80]: Again, it looks like you could simply tell struct to unpack 9 byte values from the stream. Then you wouldn't need the loop.

File: test_bt_callback.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import time
```
Comment (lumagi) [inline test_bt_callback.py:1]: This is nice, thanks for creating the scripts. Could you place them in a separate folder called `examples`?

**Review by lumagi — CHANGES_REQUESTED** [review]
I apologize for taking longer than anticipated to work through the code. Again, thank you very much for making the contribution!
The only major thing I wanted to ask of you is if you could write test cases for your code? If you look in the `test` folder, there are already several unit and integration tests for other parts in the code. I rely on the tests to be able to change parts of the code in the future because help me ensure that the code still does what it's supposed to even after I shift stuff around.

File: pyshimmer/bluetooth/bt_commands.py [hunk 8]
```diff
@@ -342,6 +343,35 @@ def receive(self, ser: BluetoothSerial) -> any:
...
+            [bytes  0- 5] offset bias values: 3 (x,y,z) 16-bit signed integers (big endian). 
+            [bytes  6-11] sensitivity values: 3 (x,y,z) 16-bit signed integers (big endian). 
+            [bytes 12-20] alignment matrix:  9 values    8-bit signed integers.
+    """
+
+    def __init__(self):
+        super().__init__(ALL_CALIBRATION_RESPONSE)
+
+        self._offset = 0x0
+        self._rlen = 0x54  # 72 bytes
```
[the remark this answers is not in the document]
Comment (MFernandezCarmona) [reply pyshimmer/bluetooth/bt_commands.py:365]: Yes, my bad

File: pyshimmer/dev/calibration.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+from typing import List, Dict, Tuple
+
+from pyshimmer.util import bit_is_set, fmt_hex
+from .channels import EChannelType
+
+
+class AllCalibration:
+
+    def __init__(self, reg_bin: bytes):
+        self.num_bytes = 84
```
[the remark this answers is not in the document]
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:28]: Ok

File: pyshimmer/dev/calibration.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        self.sensor_bytes = 21
+        self.num_sensors = 4
+
+        if len(reg_bin) < self.num_bytes:
+            raise ValueError('All calibration data must have length ' + self.num_bytes)
```
[the remark this answers is not in the document]
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:33]: Ok

File: pyshimmer/dev/calibration.py (hunk 1 above)
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:56]: Ok

File: pyshimmer/dev/calibration.py (hunk 2 above)
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:16]: PEP8, got it.

File: pyshimmer/dev/calibration.py (hunk 3 above)
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:58]: Ok

File: pyshimmer/dev/calibration.py (hunk 4 above)
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:57]: Done

File: pyshimmer/dev/calibration.py (hunk 5 above)
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:61]: Changed

File: pyshimmer/dev/calibration.py (hunk 6 above)
Comment (MFernandezCarmona) [reply pyshimmer/dev/calibration.py:80]: Ok

File: test_bt_callback.py (hunk 7 above)
Comment (MFernandezCarmona) [reply test_bt_callback.py:1]: Moved

**Comment by MFernandezCarmona** [comment]
Hi,
I think I've covered all the changes. Please have a look anyway, I'm not 100% sure I did the PEP8 formatting right (I'm using autopep8 with vscode).
Cheers

**Pull request merged**